### PR TITLE
Comments: show loading placeholder

### PR DESCRIPTION
**Pull request by gwwar**
This PR adds a new selector so we can track when we have any site comments. This allows us to know when to show a placeholder for comments management.

### Testing Instructions
- Navigate to http://calypso.localhost:3000/comments/all
- We should see a placeholder
- The placeholder should be removed to show empty content or a list of comments.

cc @Automattic/lannister

### CODE REVIEW

File: client/state/comments/reducer.js [hunk 1]
```diff
@@ -126,7 +127,24 @@ export function totalCommentsCount( state = {}, action ) {
...
+ * @param {object} action - redux action
+ * @returns {boolean} new redux state
+ */
+export function isCommentListLoading( state = false, action ) {
+	switch ( action.type ) {
+		case COMMENTS_LIST_REQUEST:
+			return true;
+		case COMMENTS_RECEIVE: //we may need a more specific success/error action here
+			return false;
+	}
```
Comment (dmsnell) [inline client/state/comments/reducer.js]: we're going to be doing lots of comment tracking, possible for different sites than are currently selected.

if we're _going_ to track this here then maybe we can think of a different approach which can give us more reliability.

what if we stored this per-site? that would prevent crossing state for the wrong sites

what if we tracked a count? that would prevent assuming the loading state finishes when we have active requests still

```js
export const siteOpenRequests = ( state = 0, { type } ) => get( [
	[ COMMENTS_LIST_REQUEST ]: state + 1,
	[ COMMENTS_RECEIVE ]: state - 1,
], type, state );

export const openRequests = keyedReducer( 'siteId', openRequests )

const hasOpenRequests = commentState => commentState.openRequests[ siteId ] > 0
```

Comment (dmsnell) [inline client/state/comments/reducer.js]: but I should not that in the end, after several iterations, I would love to remove this state entirely :winK:

Comment (gwwar) [inline client/state/comments/reducer.js]: Ideally, I'd like to show a loader with these conditions:
1. site is not available (handled as a ui check)
2. no comments are currently available for the view (eg it's ok to render stale data) AND we're making some effort to get this

I'm pretty open to ideas for figuring out how we keep that that we're making some sort of effort, and I'd be happy to try a few things, since I'm not sure we have a pattern for this yet for actions that use the data-layer.

@dmsnell So with the counter we'll want to also know when the request has failed, but I know you quite dislike the extra `_SUCCESS` and `_ERROR` actions. Without them though we can get into a state where one of the requests fails, and the others succeed so we never get back down to 0. Or similarly actions may fire for `COMMENTS_RECEIVE` in response to a post comments request, leaving us with a negative value.

We can certainly track this per site too, I was just being lazy at EOD to get something working.

Comment (gwwar) [inline client/state/comments/reducer.js]: Anyway I'll give it some more 🤔 to see if we can handle this in a more generic way.

Comment (dmsnell) [inline client/state/comments/reducer.js]: yeah @gwwar I'm not feeling particularly strong about this.

the fact that we assume requests never fail is an entirely different problem that should be fixed and is a real anti-pattern. if we weren't reusing old code in the data layer comments stuff this wouldn't be a problem.

I just like rethinking about the problem and hopping that some of this repetition will positively influence the way we build new things and enhance old ones.

Comment (gwwar) [inline client/state/comments/reducer.js]: @dmsnell I mostly want us to be unblocked for user testing this week, and I don't have a ton of uninterrupted time to think about a generic fix in the time frame. I propose that we use the old pattern for now and revisit this in a refactor soon, so this can be addressed appropriately.

This would involve dispatching `COMMENTS_LIST_SUCCESS` and `COMMENTS_LIST_ERROR` and storing an isRequesting boolean per comment list site request. 

cc @Copons @rodrigoi if folks had better ideas.

Comment (dmsnell) [inline client/state/comments/reducer.js]: @gwwar I'd be strongly opposed to adding a _new_ `SUCCESS` and `_ERROR` message in the data layer. I'm fine with pushing out a pragmatic step, but I really don't want more of those because they hurt us later on.

In these cases where "we just need it" why not just store an extra state value as some kind of enum (string is fine):

```js
import { getData, getError } from 'whatever/util/has/it';

const listState = ( state = null, action ) => {
	if ( COMMENTS_LIST_REQUEST !== action.type ) {
		return state;
	}

	if ( hasData( action ) ) {
		return 'failure';
	}

	if ( hasError( action ) ) {
		return 'loaded';
	}

	return 'loading';
}
```

this is a separate piece of state which we are using to visually indicate something in the UI and it needs no additional action types.

Comment (gwwar) [inline client/state/comments/reducer.js]: @dmsnell I'm not sure I follow here. Why would the `COMMENTS_LIST_REQUEST` action have related error data? Unless you meant something else, this still has the same issues as the counter example.

>  I'd be strongly opposed to adding a new SUCCESS and _ERROR

I understand your feelings on this 😄, but currently I don't see a good alternative here.

**Comment by Copons** [comment]
UI-wise got no complains but a doubt: is there a reason why the placeholder is outside the CSS transition group?
I've tried moving things around and it actually doesn't seem to crossfade even when inside it, but I'd still rather keep both placeholder and empty content in, and iron out the animation kinks later.

```js
<ReactCSSTransitionGroup
	className="comment-list__transition-wrapper"
	component="div"
	transitionEnterTimeout={ 300 }
	transitionLeaveTimeout={ 150 }
	transitionName="comment-list__transition"
>
	{ this.getPlaceholderOrEmptyContent() }
</ReactCSSTransitionGroup>
```

**Comment by gwwar** [comment]
> is there a reason why the placeholder is outside the CSS transition group?

I was hitting some graphical glitches, though we can add that as the next clean up task

File: client/state/comments/reducer.js (hunk 1 above)
Comment (gwwar) [inline client/state/comments/reducer.js]: Follow up note from a slack chat, `COMMENTS_LIST_REQUEST` has additional network meta attached.

Comment (gwwar) [inline client/state/comments/reducer.js]: Update: we never pass along the actions with network meta. https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom-api-middleware.js#L29

**Comment by gwwar** [comment]
Alright @dmsnell and @Copons I think this new selector should do the trick. Note that we may have a few corner cases when switching between different filter types, but we can address that in the future.

File: client/state/selectors/has-site-comments.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * Returns true if we have site comments
+ *
+ * @param  {Object}  state       Global state tree
+ * @param  {Number}  siteId      The ID of the site we're querying
+ */
+export const hasSiteComments = createSelector(
+	( state, siteId ) => {
+		const comments = get( state, 'comments.items', {} );
+		const siteKeys = Object.keys( comments )
+			.filter( key => parseInt( key.split( '-', 1 ), 10 ) === siteId );
```
Comment (gwwar) [inline client/state/selectors/has-site-comments.js]: Just a sec, I can do an early return instead of filtering through all items

**Comment by gwwar** [comment]
Pinging @roundhill @Copons @dmsnell for another 👀

File: client/my-sites/comments/comment-list/index.jsx [hunk 3]
```diff
@@ -206,6 +207,29 @@ export class CommentList extends Component {
...
+					illustration="/calypso/images/comments/illustration_comments_gray.svg"
+					illustrationWidth={ 150 }
+					key="comment-list-empty"
+					line={ emptyMessageLine }
+					title={ emptyMessageTitle }
+				/>
+			);
+		}
+		return null;
+	}
```
Comment (dmsnell) [inline client/my-sites/comments/comment-list/index.jsx]: seems like we're not adding much by extracting this, maybe adding some awkwardness with the `null`

we're basically trapping on two values: do we have comments? should we?

```js
render() {
	const { comments } = this.props;
	…
	return {
		…
		{ zeroComments && hasLoaded && (
			<CommentDetailPlaceholder … />
		) }
		{ zeroComments && ! hasLoaded && (
			<EmptyContent
				…
			/>
		) }
		…
	};
}
```

thoughts? did you find much gain by pulling it out? it seems to hide in the main render area what goes there

File: client/state/selectors/has-site-comments.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * Returns true if we have site comments
+ *
+ * @param  {Object}  state       Global state tree
+ * @param  {Number}  siteId      The ID of the site we're querying
+ */
+export const hasSiteComments = createSelector(
+	( state, siteId ) => {
+		const comments = get( state, 'comments.items', {} );
+		return !! find( Object.keys( comments ), key => parseInt( key.split( '-', 1 ), 10 ) === siteId );
```
Comment (dmsnell) [inline client/state/selectors/has-site-comments.js]: the idiom you may be looking for here is `some()`

`some()` return a boolean indicating if at least one item in the collection matches the predicate

File: client/state/selectors/has-site-comments.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	},
+	state => [ state.comments.items ]
```
Comment (dmsnell) [inline client/state/selectors/has-site-comments.js]: when I dug into `createSelector` I saw it was doing a shallow-comparison of this but I don't know if we need to do a property-wise comparison. seems like just comparing the reference equality should be enough, no?

that is to say, if we're wanting to memoize this selector, aren't we doing it wrong since Redux should produce a new top-level object when `items` changes? this code will iterate over the new and old items object on every call, which could stack up if we have lots of loaded comments.

not sure if it's a performance issue, but I can't see how adding this dependency is actually helping us

Comment (gwwar) [inline client/state/selectors/has-site-comments.js]: Reference equality should be enough, was there some other reason that I was forgetting @aduth ?

Comment (gwwar) [inline client/state/selectors/has-site-comments.js]: Regardless if we have an improvement for createSelector, that should go into it's own PR.

File: client/my-sites/comments/comment-list/index.jsx (hunk 3 above)
Comment (gwwar) [inline client/my-sites/comments/comment-list/index.jsx]: It was originally much longer with the transition group. I'd prefer to keep it here to avoid rendering falsey values. If we're rather particular about keeping it inline, I can create two bools, showPlaceholder, and showEmptyContent that can be passed through connect.

Comment (gwwar) [inline client/my-sites/comments/comment-list/index.jsx]: updated in dddd58a

File: client/state/selectors/has-site-comments.js (hunk 4 above)
Comment (gwwar) [inline client/state/selectors/has-site-comments.js]: Thanks updated in dddd58a

File: client/state/selectors/has-site-comments.js (hunk 5 above)
Comment (gwwar) [inline client/state/selectors/has-site-comments.js]: Oh I misread the code. The shallow equals is for the list of dependents rather for a single state object. So we should be OK on reference equality.

> this code will iterate over the new and old items object on every call,

I'm not sure what you mean for this scenario.

Comment (dmsnell) [inline client/state/selectors/has-site-comments.js]: please disregard. I descended one too many levels deep in the list

**Comment by gwwar** [comment]
Thanks for the reviews!
